### PR TITLE
pearson pairwise and some fast pre processing utils

### PR DESCRIPTION
**Pull request by rjenc29**
Work in progress, in my repo - enough to hopefully get the idea.  Let's catch up, ta!

### CODE REVIEW

**Comment by rjenc29** [comment]
Updated this stuff a little - I think it's nearer to being ready to PR into the main repo.  I've replaced the brute force pearson with a call to pearson_pariwise as it's just a special case, in some ways.  The included standard scale is probably generally handy and is much faster than the sklearn equiv, plus you can control ddof (which sklearn doesn't let you do).
I changed the imports to avoid overriding builtins.
Let me know is you spot any other gremlins / are OK for me to PR into the main repo for prime time.

File: fastats/maths/correlation/pearson.py [hunk 1]
```diff
@@ -25,21 +27,23 @@ def pearson(x, y):
...
+    assert A.ndim > 1
+    assert A.shape[1] > 1
 
-    first = n * np.sum(x2) - sum_x ** 2
-    second = n * np.sum(y2) - sum_y ** 2
-    denom = np.sqrt(first * second)
-    return numer / denom
+    n = A.shape[0]
+    A_std = standard_scale(A)
+    return (A_std.T @ A_std) / n
```
Comment (dwillmer) [inline fastats/maths/correlation/pearson.py]: Yea, i like the `pearson` vs `pearson_pairwise` split.

File: fastats/utilities/pre_processing.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+from numpy import empty_like, mean, std, sqrt
+from numpy import min as np_min
+from numpy import max as np_max
+
+
+def standard_scale(data, ddof=0):
```
Comment (dwillmer) [inline fastats/utilities/pre_processing.py]: Nice addition with the scaling functions - my feeling is that scaling functions fit perfectly within the fastats substitution concept, so people may want to swap them out with `my_func(..., scale=standard)` or `my_func(..., scale=min_max)` - what do you think?

File: tests/maths/correlation/test_pearson.py [hunk 3]
```diff
@@ -51,6 +52,17 @@ def test_pearson_nan_result():
     assert pearson(x, y) == approx(0.6324555320)
 
 
+@mark.parametrize('A', [load_iris().data, load_diabetes().data], ids=['iris', 'diabetes'])
```
Comment (dwillmer) [inline tests/maths/correlation/test_pearson.py]: Nice!

**Review by dwillmer — COMMENTED** [review]
Code-wise it's awesome - the only suggestion i have is to move the scaler code into something like `fastats.scaling` and have `scale`, `standard` and `min_max` functions, where `scale` is a dummy that can be used as a placeholder.

That way we can swap it out using the ast transforms quite nicely, ie `scale=standard)`.

What do you reckon?

File: tests/maths/correlation/test_pearson.py (hunk 3 above)
Comment (dwillmer) [reply tests/maths/correlation/test_pearson.py]: If we're going to `parametrize` this, should we add in some of the other sklearn datasets as well?

**Comment by rjenc29** [comment]
Morning, yep - all good comments, ta!

**Comment by rjenc29** [comment]
I added spearman pairwise - it has a nice symmetry with pearson, what do you think?

**Comment by rjenc29** [comment]
I also generalised the sklearn datasets thing - that's a great idea, it works well

**Comment by rjenc29** [comment]
I agree with making the scaling thing pluggable - haven't had time to do that yet, so will try to get it done over the weekend.  The last commit at least gets all the pairwise correlation stuff 'working' - feedback as always most welcome - early morning coding always prone to gremlins!

**Pull request merged**